src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp

```diff
@@ -0,0 +1,317 @@
+        size_t next_child = 0;
+    };
+
+    using StackWithParent = std::vector<FrameWithParent>;
+
+    QueryPlan::Node * root = nullptr;
+    StackWithParent stack;
+
+public:
+    explicit QueryPlanVisitor(QueryPlan::Node * root_) : root(root_) { }
+
+    void visit()
+    {
+        stack.push_back({.node = root});
+
+        while (!stack.empty())
+        {
+            auto & frame = stack.back();
+
+            QueryPlan::Node * current_node = frame.node;
+            QueryPlan::Node * parent_node = frame.parent_node;
+
+            logStep("back", current_node);
+
+            /// top-down visit
+            if (0 == frame.next_child)
+            {
+                logStep("top-down", current_node);
+                if (!visitTopDown(current_node, parent_node))
+                    continue;
+            }
+            /// Traverse all children
+            if (frame.next_child < frame.node->children.size())
+            {
+                auto next_frame = FrameWithParent{.node = current_node->children[frame.next_child], .parent_node = current_node};
+                ++frame.next_child;
+                logStep("push", next_frame.node);
+                stack.push_back(next_frame);
+                continue;
+            }
+
+            /// bottom-up visit
+            logStep("bottom-up", current_node);
+            visitBottomUp(current_node, parent_node);
+
+            logStep("pop", current_node);
+            stack.pop_back();
+        }
+    }
+
+    bool visitTopDown(QueryPlan::Node * current_node, QueryPlan::Node * parent_node)
+    {
+        return getDerived().visitTopDown(current_node, parent_node);
+    }
+    void visitBottomUp(QueryPlan::Node * current_node, QueryPlan::Node * parent_node)
+    {
+        getDerived().visitBottomUp(current_node, parent_node);
+    }
+
+private:
+    Derived & getDerived() { return *static_cast<Derived *>(this); }
+
+    const Derived & getDerived() const { return *static_cast<Derived *>(this); }
+
+protected:
+    void logStep(const char * prefix, const QueryPlan::Node * node)
+    {
+        if constexpr (debug_logging)
+        {
+            IQueryPlanStep * current_step = node->step.get();
+            LOG_DEBUG(
+                &Poco::Logger::get("QueryPlanVisitor"),
+                "{}: {}: {}",
+                prefix,
+                current_step->getName(),
+                reinterpret_cast<void *>(current_step));
+        }
+    }
+};
+
+constexpr bool debug_logging_enabled = true;
+
+class IsQueryDistributed : public QueryPlanVisitor<IsQueryDistributed, debug_logging_enabled>
+{
+public:
+    explicit IsQueryDistributed(QueryPlan::Node * root_) : QueryPlanVisitor<IsQueryDistributed, debug_logging_enabled>(root_) { }
+
+    explicit operator bool() const { return distributed_query; }
+
+    static bool visitTopDown(QueryPlan::Node *, QueryPlan::Node *) { return true; }
+
+    void visitBottomUp(QueryPlan::Node * current_node, QueryPlan::Node *)
```

Comment:
I have a warning `Function 'visitBottomUp' hides a non-virtual function from class 'QueryPlanVisitor<IsQueryDistributed, 1>'`. It's not very important, but we can consider renaming in QueryPlanVisitor